diff -rupN a/mozilla/nsprpub/pr/include/md/_pth.h b/mozilla/nsprpub/pr/include/md/_pth.h
--- a/mozilla/nsprpub/pr/include/md/_pth.h	2011-10-06 10:44:33.140000000 -0400
+++ b/mozilla/nsprpub/pr/include/md/_pth.h	2011-10-06 10:46:47.000000000 -0400
@@ -152,7 +152,8 @@
 	|| defined(UNIXWARE) || defined(RISCOS)	|| defined(SYMBIAN)
 #ifdef __GNU__
 /* Hurd pthreads don't have an invalid value for pthread_t. -- rmh */
-#error Using Hurd pthreads
+//This may be fixed
+//#error Using Hurd pthreads
 #endif
 #define _PT_PTHREAD_INVALIDATE_THR_HANDLE(t)  (t) = 0
 #define _PT_PTHREAD_THR_HANDLE_IS_INVALID(t)  (t) == 0
