diff --git a/sysdeps/mach/hurd/kernel-posix-cpu-timers.h b/sysdeps/mach/hurd/kernel-posix-cpu-timers.h
new file mode 100644
index 0000000..86a568f
--- /dev/null
+++ b/sysdeps/mach/hurd/kernel-posix-cpu-timers.h
@@ -0,0 +1,4 @@
+/* Parameters for the Hurd ABI for CPU clocks.  */
+
+#define CPUCLOCK_WHICH(clock)	((clock) & (clockid_t) CPUCLOCK_CLOCK_MASK)
+#define CPUCLOCK_CLOCK_MASK	3
diff --git a/sysdeps/mach/hurd/clock_settime.c b/sysdeps/mach/hurd/clock_settime.c
new file mode 100644
index 0000000..86a568f
--- /dev/null
+++ b/sysdeps/mach/hurd/clock_settime.c
@@ -0,0 +1,2 @@
+#include "kernel-posix-cpu-timers.h"
+#include <sysdeps/unix/clock_settime.c>
