diff -ur source/pacman-3.3.3/configure.ac build/pacman-3.3.3/configure.ac
--- source/pacman-3.3.3/configure.ac	2009-11-10 18:12:11.000000000 +0100
+++ build/pacman-3.3.3/configure.ac	2010-03-07 15:49:49.009375795 +0100
@@ -142,8 +142,8 @@
 	AC_MSG_RESULT(yes)
 	AC_DEFINE([INTERNAL_DOWNLOAD], , [Use internal download library])
 	# Check for a download library if it was actually requested
-	AC_CHECK_LIB([fetch], [fetchParseURL], ,
-		AC_MSG_ERROR([libfetch is needed to compile with internal download support]) )
+	AC_CHECK_LIB([fetch], [fetchParseURL], [AC_SUBST([LIBS], ["-lfetch -lssl -lcrypto -ldl $LIBS"])],
+		AC_MSG_ERROR([libfetch is needed to compile with internal download support]), [-lssl -lcrypto -ldl] )
 	# Check if libfetch supports conditional GET
 	# (version >=2.21, struct url has member last_modified)
 	AC_CHECK_MEMBER(struct url.last_modified, ,
@@ -296,10 +296,10 @@
 fi
 
 # Enable or disable use of git version in pacman version string
-AC_CHECK_PROGS([GIT], [git])
-AC_CHECK_FILE([.git/], hasgitdir=yes)
 AC_MSG_CHECKING(whether to use git version if available)
 if test "x$wantgitver" = "xyes" ; then
+	AC_CHECK_PROGS([GIT], [git])
+	AC_CHECK_FILE([.git/], hasgitdir=yes)
 	if test $GIT -a "x$hasgitdir" = "xyes"; then
 		AC_MSG_RESULT([yes])
 		usegitver=yes
diff -ur source/pacman-3.3.3/lib/libalpm/alpm.h build/pacman-3.3.3/lib/libalpm/alpm.h
--- source/pacman-3.3.3/lib/libalpm/alpm.h	2009-09-30 03:38:36.000000000 +0200
+++ build/pacman-3.3.3/lib/libalpm/alpm.h	2010-03-07 15:18:11.869981050 +0100
@@ -35,6 +35,8 @@
 
 #define DEPRECATED __attribute__((deprecated))
 
+#define PATH_MAX	4096
+
 /*
  * Arch Linux Package Management library
  */
