diff -upr Mesa-7.8.1.orig//src/gallium/auxiliary/os/os_time.c Mesa-7.8.1/src/gallium/auxiliary/os/os_time.c
--- Mesa-7.8.1.orig//src/gallium/auxiliary/os/os_time.c	2010-04-25 09:29:05.000000000 +0000
+++ Mesa-7.8.1/src/gallium/auxiliary/os/os_time.c	2010-04-25 09:33:15.000000000 +0000
@@ -37,9 +37,9 @@
 
 #if !defined(PIPE_OS_EMBEDDED)
 
-#if defined(PIPE_OS_LINUX) || defined(PIPE_OS_BSD) || defined(PIPE_OS_SOLARIS) || defined(PIPE_OS_APPLE) || defined(PIPE_OS_HAIKU)
+/*#if defined(PIPE_OS_LINUX) || defined(PIPE_OS_BSD) || defined(PIPE_OS_SOLARIS) || defined(PIPE_OS_APPLE) || defined(PIPE_OS_HAIKU) */
 #  include <sys/time.h> /* timeval */
-#elif defined(PIPE_SUBSYSTEM_WINDOWS_DISPLAY)
+/*#elif defined(PIPE_SUBSYSTEM_WINDOWS_DISPLAY)
 #  include <windows.h>
 #  include <winddi.h>
 #elif defined(PIPE_SUBSYSTEM_WINDOWS_MINIPORT)
@@ -49,7 +49,7 @@ extern VOID KeQuerySystemTime(PLARGE_INT
 #  include <windows.h>
 #else
 #  error Unsupported OS
-#endif
+#endif */
 
 #include "os_time.h"
 
