diff -rup0 orig//Modules/FindGTK2.cmake new//Modules/FindGTK2.cmake
--- orig//Modules/FindGTK2.cmake	2010-06-28 15:15:57.000000000 +0000
+++ new//Modules/FindGTK2.cmake	2010-07-03 20:13:18.000000000 +0000
@@ -441,0 +442 @@ foreach(_GTK2_component ${GTK2_FIND_COMP
+        _GTK2_FIND_INCLUDE_DIR(GTK2_PANGOMMCONFIG_INCLUDE_DIR pangommconfig.h)
diff -rup0 orig//Modules/Platform/GNU.cmake new//Modules/Platform/GNU.cmake
--- orig//Modules/Platform/GNU.cmake	2010-06-28 15:15:57.000000000 +0000
+++ new//Modules/Platform/GNU.cmake	2010-07-03 20:12:49.000000000 +0000
@@ -6,0 +7 @@ SET(CMAKE_SHARED_LIBRARY_RUNTIME_C_FLAG_
+SET(CMAKE_SHARED_LIBRARY_RPATH_LINK_C_FLAG "-Wl,-rpath-link,")
@@ -7,0 +9,4 @@ SET(CMAKE_SHARED_LIBRARY_SONAME_C_FLAG "
+SET(CMAKE_SHARED_LIBRARY_SONAME_Fortran_FLAG "-Wl,-soname,")
+SET(CMAKE_EXE_EXPORTS_C_FLAG "-Wl,--export-dynamic")
+SET(CMAKE_EXE_EXPORTS_CXX_FLAG "-Wl,--export-dynamic")
+
diff -rup0 orig//Utilities/cmlibarchive/libarchive/archive_read_disk_entry_from_file.c new//Utilities/cmlibarchive/libarchive/archive_read_disk_entry_from_file.c
--- orig//Utilities/cmlibarchive/libarchive/archive_read_disk_entry_from_file.c	2010-06-28 15:16:06.000000000 +0000
+++ new//Utilities/cmlibarchive/libarchive/archive_read_disk_entry_from_file.c	2010-07-03 19:44:07.000000000 +0000
@@ -63,0 +64,4 @@ __FBSDID("$FreeBSD$");
+#ifndef PATH_MAX
+#define PATH_MAX 4096
+#endif
+
diff -rup0 orig//Utilities/cmlibarchive/libarchive/archive_write_disk.c new//Utilities/cmlibarchive/libarchive/archive_write_disk.c
--- orig//Utilities/cmlibarchive/libarchive/archive_write_disk.c	2010-06-28 15:16:06.000000000 +0000
+++ new//Utilities/cmlibarchive/libarchive/archive_write_disk.c	2010-07-03 19:44:35.000000000 +0000
@@ -98,0 +99,4 @@ __FBSDID("$FreeBSD: src/lib/libarchive/a
+#ifndef PATH_MAX
+#define PATH_MAX 4096
+#endif
+
