diff -rupN namcap-2.6/Namcap/directoryname.py namcap-2.6-nousr/Namcap/directoryname.py
--- namcap-2.6/Namcap/directoryname.py	2010-06-08 04:43:19.000000000 +0000
+++ namcap-2.6-nousr/Namcap/directoryname.py	2010-08-25 16:56:13.600000000 +0000
@@ -25,7 +25,7 @@ class package:
 	def prereq(self):
 		return "tar"
 	def analyze(self, pkginfo, tar):
-		valid_paths = ['bin/', 'etc/', 'usr/bin/', 'usr/sbin/', 'usr/lib', 'usr/include/', 'usr/share/', 'opt/', '.PKGINFO', '.INSTALL', '.FILELIST', '._install', 'lib/', 'sbin/', 'srv/', 'var/lib/', 'var/opt/', 'var/spool/', 'var/lock/', 'var/state/', 'var/run/', 'var/log/', '.CHANGELOG']
+		valid_paths = ['bin/', 'etc/', 'sbin/', 'lib/', 'include/', 'share/', 'opt/', '.PKGINFO', '.INSTALL', '.FILELIST', '._install', 'srv/', 'var/lib/', 'var/opt/', 'var/spool/', 'var/lock/', 'var/state/', 'var/run/', 'var/log/', '.CHANGELOG']
 		ret = [[], [], []]
 		for i in tar.getnames():
 			# Replace multiple /'s at the end of a string with a single /
diff -rupN namcap-2.6/Namcap/elffiles.py namcap-2.6-nousr/Namcap/elffiles.py
--- namcap-2.6/Namcap/elffiles.py	2010-06-08 04:40:15.000000000 +0000
+++ namcap-2.6-nousr/Namcap/elffiles.py	2010-08-25 16:56:49.870000000 +0000
@@ -24,7 +24,7 @@ def scanelf(invalid_elffiles, dirname, n
 	'''Method to scan for ELF files in invalid directories'''
 
 	# Valid directories for ELF files
-	valid_dirs = ['/bin/', '/sbin/', '/usr/bin/', '/usr/sbin/', '/lib/', '/usr/lib/']	
+	valid_dirs = ['/bin/', '/sbin/', '/lib/']	
 	
 	for i in names:
 		file_path = os.path.join(dirname, i)
diff -rupN namcap-2.6/Namcap/fhsinfopages.py namcap-2.6-nousr/Namcap/fhsinfopages.py
--- namcap-2.6/Namcap/fhsinfopages.py	2010-06-08 04:47:52.000000000 +0000
+++ namcap-2.6-nousr/Namcap/fhsinfopages.py	2010-08-25 16:57:11.000000000 +0000
@@ -29,9 +29,9 @@ class package:
 	def analyze(self, pkginfo, tar):
 		ret = [[], [], []]
 		for i in tar.getmembers():
-			if i.name.startswith('usr/info'):
+			if i.name.startswith('info'):
 				ret[0].append(("non-fhs-info-page %s", i.name))
-			elif not i.name.startswith('usr/share/info'):
+			elif not i.name.startswith('share/info'):
 				for part in i.name.split(os.sep):
 					if part == "info":
 						ret[1].append(("potential-non-fhs-info-page %s", i.name))
diff -rupN namcap-2.6/Namcap/fhsmanpages.py namcap-2.6-nousr/Namcap/fhsmanpages.py
--- namcap-2.6/Namcap/fhsmanpages.py	2010-06-08 04:38:41.000000000 +0000
+++ namcap-2.6-nousr/Namcap/fhsmanpages.py	2010-08-25 16:57:23.000000000 +0000
@@ -28,8 +28,8 @@ class package:
 	def prereq(self):
 		return "tar"
 	def analyze(self, pkginfo, tar):
-		gooddir = 'usr/share/man'
-		bad_dir = 'usr/man'
+		gooddir = 'share/man'
+		bad_dir = 'man'
 		ret = [[], [], []]
 		for i in tar.getmembers():
 			if i.name.startswith(bad_dir):
diff -rupN namcap-2.6/Namcap/gnomemime.py namcap-2.6-nousr/Namcap/gnomemime.py
--- namcap-2.6/Namcap/gnomemime.py	2010-06-08 04:51:17.000000000 +0000
+++ namcap-2.6-nousr/Namcap/gnomemime.py	2010-08-25 16:57:51.000000000 +0000
@@ -26,12 +26,12 @@ class package:
         return "tar"
     def analyze(self, pkginfo, tar):
         mime_files = [
-                'usr/share/applications/mimeinfo.cache',
-                'usr/share/mime/XMLnamespaces', 
-                'usr/share/mime/aliases', 
-                'usr/share/mime/globs', 
-                'usr/share/mime/magic', 
-                'usr/share/mime/subclasses'
+                'share/applications/mimeinfo.cache',
+                'share/mime/XMLnamespaces', 
+                'share/mime/aliases', 
+                'share/mime/globs', 
+                'share/mime/magic', 
+                'share/mime/subclasses'
                 ]
 
         ret = [[], [], []]
diff -rupN namcap-2.6/Namcap/hicoloricons.py namcap-2.6-nousr/Namcap/hicoloricons.py
--- namcap-2.6/Namcap/hicoloricons.py	2010-06-08 04:49:17.000000000 +0000
+++ namcap-2.6-nousr/Namcap/hicoloricons.py	2010-08-25 16:58:02.630000000 +0000
@@ -26,7 +26,7 @@ class package:
 		return "tar"
 	def analyze(self, pkginfo, tar):
 		ret = [[], [], []]
-		if "usr/share/icons/hicolor" in tar.getnames():
+		if "share/icons/hicolor" in tar.getnames():
 			if hasattr(pkginfo, "depends"):
 				if "hicolor-icon-theme" not in pkginfo.depends:
 					ret[0].append(("dependency-detected-not-included %s", ("hicolor-icon-theme",)))
diff -rupN namcap-2.6/Namcap/infodirectory.py namcap-2.6-nousr/Namcap/infodirectory.py
--- namcap-2.6/Namcap/infodirectory.py	2010-06-08 04:46:48.000000000 +0000
+++ namcap-2.6-nousr/Namcap/infodirectory.py	2010-08-25 16:58:29.190000000 +0000
@@ -27,7 +27,7 @@ class package:
 	def analyze(self, pkginfo, tar):
 		ret = [[], [], []]
 		for i in tar.getnames():
-			if i == "usr/share/info/dir":
+			if i == "share/info/dir":
 				ret[0].append(("info-dir-file-present %s", i))
 		return ret
 	def type(self):
diff -rupN namcap-2.6/Namcap/licensepkg.py namcap-2.6-nousr/Namcap/licensepkg.py
--- namcap-2.6/Namcap/licensepkg.py	2010-06-08 04:45:30.000000000 +0000
+++ namcap-2.6-nousr/Namcap/licensepkg.py	2010-08-25 16:59:05.000000000 +0000
@@ -31,7 +31,7 @@ class package:
 		if not hasattr(pkginfo, 'license') or len(pkginfo.license) == 0:
 			ret[0].append(("missing-license", ()))
 		else:
-			licensepaths = [x for x in tar.getnames() if x.startswith('usr/share/licenses') and not x.endswith('/')]
+			licensepaths = [x for x in tar.getnames() if x.startswith('share/licenses') and not x.endswith('/')]
 			licensedirs = [os.path.split(os.path.split(x)[0])[1] for x in licensepaths]
 			licensefiles = [os.path.split(x)[1] for x in licensepaths]
 			# Check all licenses for validity
@@ -39,12 +39,12 @@ class package:
 				lowerlicense = license.lower()
 				if lowerlicense.startswith('custom') or lowerlicense in ("bsd", "mit", "isc", "python", "zlib", "libpng"):
 					if pkginfo.name not in licensedirs:
-						ret[0].append(("missing-custom-license-dir usr/share/licenses/%s", pkginfo.name))
+						ret[0].append(("missing-custom-license-dir share/licenses/%s", pkginfo.name))
 					elif len(licensefiles) == 0:
-						ret[0].append(("missing-custom-license-file usr/share/licenses/%s/*", pkginfo.name))
+						ret[0].append(("missing-custom-license-file share/licenses/%s/*", pkginfo.name))
 				# A common license
 				else:
-					commonlicenses = [x.lower() for x in os.listdir('/usr/share/licenses/common')]
+					commonlicenses = [x.lower() for x in os.listdir('/share/licenses/common')]
 					if lowerlicense not in commonlicenses:
 						ret[0].append(("not-a-common-license %s", license))
 		return ret
diff -rupN namcap-2.6/Namcap/lotsofdocs.py namcap-2.6-nousr/Namcap/lotsofdocs.py
--- namcap-2.6/Namcap/lotsofdocs.py	2010-06-08 04:46:17.000000000 +0000
+++ namcap-2.6-nousr/Namcap/lotsofdocs.py	2010-08-25 16:59:38.000000000 +0000
@@ -29,7 +29,7 @@ class package:
 		if hasattr(pkginfo, 'name'):
 			if pkginfo.name.endswith('-doc'):
 				return ret
-		docdir = 'usr/share/doc'
+		docdir = 'share/doc'
 		size = 0
 		docsize = 0
 
diff -rupN namcap-2.6/Namcap/mimefiles.py namcap-2.6-nousr/Namcap/mimefiles.py
--- namcap-2.6/Namcap/mimefiles.py	2010-06-08 04:44:59.000000000 +0000
+++ namcap-2.6-nousr/Namcap/mimefiles.py	2010-08-25 17:01:29.850000000 +0000
@@ -21,12 +21,12 @@ class package:
     def short_name(self):
         return "mimefiles"
     def long_name(self):
-        return "Check for files in /usr/share/mime"
+        return "Check for files in /share/mime"
     def prereq(self):
         return "tar"
     def analyze(self, pkginfo, tar):        
         ret = [[], [], []]
-        if 'usr/share/mime' in tar.getnames():            
+        if 'share/mime' in tar.getnames():            
             if hasattr(pkginfo, "depends"):
                 if "shared-mime-info" not in pkginfo.depends:
                     ret[0].append(("dependency-detected-not-included %s", ("shared-mime-info",)))
diff -rupN namcap-2.6/Namcap/rpath.py namcap-2.6-nousr/Namcap/rpath.py
--- namcap-2.6/Namcap/rpath.py	2010-06-08 04:50:58.000000000 +0000
+++ namcap-2.6-nousr/Namcap/rpath.py	2010-08-25 17:00:38.440000000 +0000
@@ -20,9 +20,9 @@
 import os, subprocess, re
 from Namcap.util import is_elf, clean_filename
 
-allowed = ['/usr/lib', '/lib', '$ORIGIN', '${ORIGIN}']
+allowed = ['/lib', '/lib', '$ORIGIN', '${ORIGIN}']
 allowed_toplevels = map(lambda s: s + '/', allowed)
-warn = ['/usr/local/lib']
+warn = ['/local/lib']
 libpath = re.compile('Library rpath: \[(.*)\]')
 
 def checkrpath(insecure_rpaths, dirname, names):
diff -rupN namcap-2.6/namcap.1 namcap-2.6-nousr/namcap.1
--- namcap-2.6/namcap.1	2010-06-14 16:45:06.000000000 +0000
+++ namcap-2.6-nousr/namcap.1	2010-08-25 17:03:54.000000000 +0000
@@ -17,7 +17,7 @@ Do not run RULELIST rules on the package
 display information messages
 .TP
 .B "\-m, \-\-machine\-readable"
-displays easily parseable namcap tags instead of the normal human readable description; for example using non-fhs-man-page instead of "Non-FHS man page (%s) found. Use /usr/share/man instead". A full list of namcap tags along with their human readable descriptions can be found at /usr/share/namcap/tags.
+displays easily parseable namcap tags instead of the normal human readable description; for example using non-fhs-man-page instead of "Non-FHS man page (%s) found. Use /share/man instead". A full list of namcap tags along with their human readable descriptions can be found at /share/namcap/tags.
 .TP
 \fB\-r\fR RULELIST, \fB\-\-rules=\fRRULELIST
 only apply RULELIST rules to the package
@@ -49,10 +49,10 @@ This module runs ldd on all executables,
 Some cases where namcap fails are dlopen() and obscure links.  dlopen()'d libraries don't show up because they are loaded at run time: in the case of a program that loads plugins.  Obscure links are the cases where only a small portion of the package needs something to run; usually, the small portion won't be accessed unless that thing is installed (i.e. a java plugin for mozilla included in the java package).
 .TP
 .B directoryname
-Checks the directory names in the package file list for standard directories (i.e. /etc, /usr/bin, /usr/lib).  All files in directories that are non-standard are returned
+Checks the directory names in the package file list for standard directories (i.e. /etc, /bin, /lib).  All files in directories that are non-standard are returned
 .TP
 .B elffiles
-Gives an error if it finds ELF (binary) files outside the standard paths (/bin, /usr/bin, etc.)
+Gives an error if it finds ELF (binary) files outside the standard paths (/bin etc.)
 .TP
 .B emptydir
 Looks for directories in a package that don't contain any files
@@ -74,10 +74,10 @@ Checks for generated GNOME mime files
 .TP
 .B hicoloricons
 Checks whether the hicolor icon cache is updated when a package installs
-files in /usr/share/icons/hicolor
+files in /share/icons/hicolor
 .TP
 .B infodirectory
-Verifies that a package does not contain the info directory file (/usr/share/info/dir)
+Verifies that a package does not contain the info directory file (/share/info/dir)
 .TP
 .B invalidstartdir
 Any PKGBUILDs that refer to files in $startdir are incorrect
@@ -91,10 +91,10 @@ Verifies that the licenses variable has
 .B licensepkg
 Verifies that the licenses variable has been filled in in a package. For packages with a custom license it
 checks whether the license file has been installed in
-/usr/share/licenses/$pkgname/
+/share/licenses/$pkgname/
 .TP
 .B mimefiles
-Checks whether update-mime-database is called when the package installs files in /usr/share/mime
+Checks whether update-mime-database is called when the package installs files in /share/mime
 .TP
 .B perllocal
 Searches for perllocal.pod.  perllocal.pod is a nasty file that's included during most perl module installations
@@ -103,7 +103,7 @@ Searches for perllocal.pod.  perllocal.p
 Checks basic file and and directory permissions.  It returns warnings about world readable bits not being set, world writable bits being set, and world executable bit not being set for directories
 .TP
 .B rpath
-Gives an error if a binary has RPATH set to something other than /usr/lib
+Gives an error if a binary has RPATH set to something other than /lib
 .TP
 .B scrollkeeper
 Verifies that there aren't any scrollkeeper directories
diff -rupN namcap-2.6/namcap.py namcap-2.6-nousr/namcap.py
--- namcap-2.6/namcap.py  2010-06-08 04:28:12.000000000 +0000
+++ namcap-2.6-nousr/namcap.py  2010-08-25 16:49:48.000000000 +0000
@@ -1,4 +1,4 @@
-#! /usr/bin/env python
+#! /bin/env python
 # 
 # namcap - A Pacman package analyzer
 # Copyright (C) 2003-2009 Jason Chu <jason@archlinux.org>
@@ -71,7 +71,7 @@ def process_tags(filename=None, machine=
 		return lambda s: s
 	tags = {}
 	if not filename:
-		filename="/usr/share/namcap/namcap-tags"
+		filename="/share/namcap/namcap-tags"
 	f = open(filename)
 	for i in f.readlines():
 		if i[0] == "#" or i.strip() == "": continue
diff -rupN namcap-2.6/namcap-tags namcap-2.6-nousr/namcap-tags
--- namcap-2.6/namcap-tags	2010-06-08 13:39:08.000000000 +0000
+++ namcap-2.6-nousr/namcap-tags	2010-08-25 17:07:25.000000000 +0000
@@ -29,7 +29,7 @@ file-referred-in-startdir :: File refere
 file-world-writable %s :: File (%s) has the world writable bit set.
 gnome-mime-file %s :: File (%s) is an auto-generated GNOME mime file
 hardlink-found %s points to %s :: Hard link (%s) found that points to %s
-hicolor-icon-cache-not-updated :: Files in /usr/share/icons/hicolor but no call to gtk-update-icon-cache or xdg-icon-resource to update the icon cache
+hicolor-icon-cache-not-updated :: Files in /share/icons/hicolor but no call to gtk-update-icon-cache or xdg-icon-resource to update the icon cache
 improper-checksum %s %s :: Improper %s: '%s'
 incorrect-library-permissions %s :: Library (%s) does not have permission set to 644 or 444
 incorrect-permissions %s (%s/%s) :: File (%s) has %s/%s permissions
@@ -39,17 +39,17 @@ libtool-file-present %s :: File (%s) is
 library-no-package-associated %s :: Referenced library '%s' is an uninstalled dependency
 link-level-dependence %s in %s :: Link-level dependence (%s) in file %s
 lots-of-docs %f :: Package was %.0f%% docs by size; maybe you should split out a docs package
-mime-cache-not-updated :: Mime-file found. Add "update-mime-database usr/share/mime" to the install file
+mime-cache-not-updated :: Mime-file found. Add "update-mime-database share/mime" to the install file
 missing-contributor :: Missing Contributor tag
-missing-custom-license-dir usr/share/licenses/%s :: Missing custom license directory (usr/share/licenses/%s)
-missing-custom-license-file usr/share/licenses/%s/* :: Missing custom license file in package (usr/share/licenses/%s/*)
+missing-custom-license-dir share/licenses/%s :: Missing custom license directory (share/licenses/%s)
+missing-custom-license-file share/licenses/%s/* :: Missing custom license file in package (share/licenses/%s/*)
 missing-license :: Missing license
 missing-maintainer :: Missing Maintainer tag
 missing-checksums :: Missing checksums
 missing-url :: Missing url
-non-fhs-info-page %s :: Non-FHS info page (%s) found. Use /usr/share/info instead
-non-fhs-man-page %s :: Non-FHS man page (%s) found. Use /usr/share/man instead
-not-a-common-license %s :: %s is not a common license (it's not in /usr/share/licenses/common/)
+non-fhs-info-page %s :: Non-FHS info page (%s) found. Use /share/info instead
+non-fhs-man-page %s :: Non-FHS man page (%s) found. Use /share/man instead
+not-a-common-license %s :: %s is not a common license (it's not in /share/licenses/common/)
 not-enough-checksums %s %i needed :: Not enough %s: %i needed
 package-name-in-uppercase :: No upper case letters in package names
 perllocal-pod-present %s :: perllocal.pod found in %s.
