diff -u source/libfetch-2.30/common.c build/libfetch-2.30/common.c
--- source/libfetch-2.30/common.c	2010-02-27 02:13:45.000000000 +0100
+++ build/libfetch-2.30/common.c	2010-03-06 22:24:09.740305036 +0100
@@ -876,17 +876,21 @@
 int
 fetch_netrc_auth(struct url *url)
 {
+#ifndef __GNU__
 	char fn[PATH_MAX];
+#endif
 	const char *word;
 	char *p;
 	FILE *f;
 
 	if ((p = getenv("NETRC")) != NULL) {
+#ifndef __GNU__
 		if (snprintf(fn, sizeof(fn), "%s", p) >= (int)sizeof(fn)) {
 			fetch_info("$NETRC specifies a file name "
 			    "longer than PATH_MAX");
 			return (-1);
 		}
+#endif
 	} else {
 		if ((p = getenv("HOME")) != NULL) {
 			struct passwd *pwd;
@@ -895,11 +899,19 @@
 			    (p = pwd->pw_dir) == NULL)
 				return (-1);
 		}
+		
+#ifndef __GNU__
 		if (snprintf(fn, sizeof(fn), "%s/.netrc", p) >= (int)sizeof(fn))
 			return (-1);
+#endif
+
 	}
 
+#ifdef __GNU__
+	if ((f = fopen(p, "r")) == NULL)
+#else
 	if ((f = fopen(fn, "r")) == NULL)
+#endif
 		return (-1);
 	while ((word = fetch_read_word(f)) != NULL) {
 		if (strcmp(word, "default") == 0)
diff -u source/libfetch-2.30/ftp.c build/libfetch-2.30/ftp.c
--- source/libfetch-2.30/ftp.c	2010-02-27 02:13:45.000000000 +0100
+++ build/libfetch-2.30/ftp.c	2010-03-06 22:45:16.499002199 +0100
@@ -98,6 +98,10 @@
 #include "common.h"
 #include "ftperr.h"
 
+#ifndef PATH_MAX
+#define PATH_MAX 4096
+#endif
+
 #define FTP_ANONYMOUS_USER	"anonymous"
 
 #define FTP_CONNECTION_ALREADY_OPEN	125
diff -u source/libfetch-2.30/http.c build/libfetch-2.30/http.c
--- source/libfetch-2.30/http.c	2010-02-27 02:13:45.000000000 +0100
+++ build/libfetch-2.30/http.c	2010-03-06 22:48:57.659396385 +0100
@@ -110,6 +110,10 @@
 #include "common.h"
 #include "httperr.h"
 
+#ifndef PATH_MAX
+#define PATH_MAX 4096
+#endif
+
 /* Maximum number of redirects to follow */
 #define MAX_REDIRECT 5
 
