--- xorg-cf-files-1.0.3/gnu.cf.orig	2010-09-08 17:28:26.750000000 +0000
+++ xorg-cf-files-1.0.3/gnu.cf	2010-09-08 17:28:42.290000000 +0000
@@ -99,7 +99,7 @@ XCOMM Not implemented and will always fa
 #define CplusplusCmd		c++
 #ifndef TermcapLibrary
 #if UseStaticTermcapLib
-#define TermcapLibrary		StaticLibrary(/usr/lib,ncurses)
+#define TermcapLibrary		StaticLibrary(/lib,ncurses)
 #else
 #define TermcapLibrary		-lncurses
 #endif

--- xorg-cf-files-1.0.3/X11.tmpl.orig	2006-07-15 10:45:23.000000000 +0000
+++ xorg-cf-files-1.0.3/X11.tmpl	2006-07-15 10:46:19.000000000 +0000
@@ -1502,10 +1502,10 @@
 #define DocPdfDir $(DOCDIR)/PDF
 #endif
 #ifndef FontDir
-#define FontDir $(LIBDIR)/fonts
+#define FontDir /share/fonts
 #endif
 #ifndef FontEncDir
-#define FontEncDir $(LIBDIR)/fonts/encodings
+#define FontEncDir /share/fonts/encodings
 #endif
 #ifndef AdmDir
 #define AdmDir /adm
